ops/ruby/devops-toolkit/tasks/dashboards.rb

```diff
@@ -0,0 +1,153 @@
+      dashboard_client.list_dashboards(env.formatted_project_number).each do |dash|
+        @logger.info("#{dash.display_name}\t#{self.class.dashboard_console_link(dash, env.project_id)}")
+        @logger.info(dash.to_json)
+        dash.grid_layout.widgets.each do |widget|
+          @logger.info("\tChart \"#{widget.title}\"")
+        end
+      end
+    end
+  end
+
+  # Take a dashboard from the source environment, tweak it for each target environment, and
+  # create a copy.
+  # TODO(jaycarlton) We have two replicate tasks already, and the pattern vis-a-vis the visitor
+  # is the same. So we can think about making this replication visiting pattern part of the visitor
+  # itself, or otherwise genericker than it is currently.
+  def replicate
+    source_dashboard = get_source_dashboard
+    @logger.info("Retrieved source dashboard #{source_dashboard.to_json}")
+    copy_to_target_envs(source_dashboard)
+  end
+
+  # Construct a link to the monitoring console. Note that in the IntelliJ debugger console,
+  # these are clickable links. I don't see this out of the box in my OS's terminal though.
+  def self.dashboard_console_link(dashboard, project_id)
+    "https://console.cloud.google.com/monitoring/dashboards/custom/#{dashboard_number(dashboard)}" +
+        "?project=#{project_id}"
+  end
+
+  # The dashboard link requires the number to the far right in the name,
+  # which  is formated like projects/<project_number>/dashboards/<dashboard_number>
+  def self.dashboard_number(dashboard)
+    dashboard.name.split('/')[-1]
+  end
+
+  # Create the resource string/URI for a dashboard. The API expects the format
+  # projects/{project_id_or_number}/dashboards/{dashboard_id}
+  def self.dashboard_resource_path(env, dashboard_number)
+    "projects/#{env.project_number}/dashboards/#{dashboard_number}"
+  end
+
+  # Return true if this dashbaord is already present in the current environment.
+  # Irritatingly, we can't just just get_dashboard for this, as it throws an exception if
+  # it's not found.
+  def self.dashboard_exists?(dashboard_client, env, dashboard_name)
+    # all = dashboard_client.list_dashboards(env.formatted_project_number, filter: CUSTOM_DASHBOARD_FILTER)
+    all = dashboard_client.list_dashboards(env.formatted_project_number)
+    all.any? { |dash| dash.name == dashboard_name }
+  end
+
+  private
+
+  def get_source_dashboard
+    source_env = @visitor.env_map[@source_env_short_name]
+    # source_env = @visitor.env_by_short_name(@source_env_short_name)
+    result = nil
+    @visitor.visit(source_env) do |env|
+      dashboard_client = Google::Cloud::Monitoring::Dashboard::V1::DashboardsServiceClient.new
+      source_dashboard_path = self.class.dashboard_resource_path(env, @source_dashboard_number)
+      result = dashboard_client.get_dashboard(source_dashboard_path)
+      @logger.info("Template dashboard: display_name: #{result.display_name}, name: #{result.name}")
+      @logger.info(result)
+    end
+    result
```

Comment:
the visitor makes it seem like we're visiting multiple environments but only the last one is being returned as part of `result`. Is it just visiting one environment every time this is invoked?